nexus/src/authz/context.rs

```diff
@@ -93,19 +93,26 @@ impl Context {
                 error
             ))),
             Ok(false) => {
-                let error = if is_authn {
-                    Error::Forbidden
-                } else {
-                    // If the user did not authenticate successfully, this will
-                    // become a 401 rather than a 403.
+                Err(if !is_authn {
```

Comment:
This is a small but important change.

Prior to this, if you failed an authz check, the logic went like this:

1. Pick a "fallback" error to signify "you didn't have permissions to do this".  This would be 401 if you were unauthenticated and 403 if you're authenticated.
2. Check with the resource impl (via the call to `on_unauthorized` at L109).  The only interesting impl is [this one](https://github.com/oxidecomputer/omicron/blob/c8d98d3593cf85d1d677b4390ac7afde6b8c50ce/nexus/src/authz/api_resources.rs#L85-L107), which is used for most API resources.  Pass the fallback error, but return whatever the resource says to return.

The intent of the `on_unauthorized` call is to determine if you're allowed to even know this resource exists.  If so, the "fallback error" (401 or 403) is fine.  If not, return a 404 instead.  The `on_unauthorized` impl I linked above is what actually does this.

After this change, the 401 always takes precedence.  If you fail the authz check and you're not authenticated, you always get a 401 -- we don't invoke `on_unauthorized()` at all.  If you're authenticated, then we do the same as before: 403 if you're allowed to read the resource, 404 otherwise.

This change isn't strictly necessary.  I did this for consistency of the public API.  If we _just_ change lookups to use `pool_authorized()`, then all the endpoints that do a lookup behave this way without this change here.  But the endpoints that don't do lookups for whatever reason (see #845) would still return 404s in some cases for unauthenticated users that fail an authz check.  I thought it better to put this logic here for consistency.  If you ever add a new endpoint (whether it hits the database or not), and the user fails an authz check, and they're not authenticated, they'll get a 401, instead of having it be dependent on how the endpoint was implemented.